### PR TITLE
TF-1080 Implement delete multiple mailbox folder

### PR DESCRIPTION
**Pull request by dab246**
#1080 

### Resolved

https://user-images.githubusercontent.com/80730648/197955299-0d89fa9f-9b2c-4552-b2ef-b1e1b9ba1480.mp4

### CODE REVIEW

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 1]
```diff
@@ -173,8 +174,10 @@ class MailboxController extends BaseMailboxController {
             _createNewMailboxSuccess(success);
           } else if (success is SearchMailboxSuccess) {
             _searchMailboxSuccess(success);
-          } else if (success is DeleteMultipleMailboxSuccess) {
-            _deleteMailboxSuccess(success);
+          } else if (success is DeleteMultipleMailboxAllSuccess) {
+            _deleteMultipleMailboxSuccess(success.listMailboxIdDeleted, success.currentMailboxState);
+          } else if (success is DeleteMultipleMailboxHasSomeSuccess) {
+            _deleteMultipleMailboxSuccess(success.listMailboxIdDeleted, success.currentMailboxState);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:180]: why not?
```suggestion
          } else if (success is DeleteMultipleMailboxAllSuccess || success is DeleteMultipleMailboxHasSomeSuccess) {
            _deleteMultipleMailboxSuccess(success.listMailboxIdDeleted, success.currentMailboxState);
          }
```

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 2]
```diff
@@ -533,29 +536,35 @@ class MailboxController extends BaseMailboxController {
...
-  void pressMailboxSelectionAction(BuildContext context, MailboxActions actions,
-      List<PresentationMailbox> selectionMailbox) {
+  void pressMailboxSelectionAction(
+      BuildContext context, 
+      MailboxActions actions,
+      List<PresentationMailbox> selectedMailboxList
+  ) {
     switch(actions) {
       case MailboxActions.delete:
-        _openConfirmationDialogDeleteMailboxAction(context, selectionMailbox.first);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:540]: :)

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 3]
```diff
@@ -533,29 +536,35 @@ class MailboxController extends BaseMailboxController {
...
+        _openConfirmationDialogDeleteMultipleMailboxAction(context, selectedMailboxList);
         break;
       case MailboxActions.rename:
-        _openDialogRenameMailboxAction(context, selectionMailbox.first);
+        _openDialogRenameMailboxAction(context, selectedMailboxList.first);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:549]: are you sure the button `rename` will be not enable when select multiple mailbox?

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 4]
```diff
@@ -533,29 +536,35 @@ class MailboxController extends BaseMailboxController {
...
         break;
       case MailboxActions.markAsRead:
-        _markAsReadMailboxAction(context, selectionMailbox.first);
+        _markAsReadMailboxAction(context, selectedMailboxList.first);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:552]: idem

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 5]
```diff
@@ -533,29 +536,35 @@ class MailboxController extends BaseMailboxController {
...
         break;
       case MailboxActions.move:
-        _moveMailboxAction(selectionMailbox.first);
+        _moveMailboxAction(selectedMailboxList.first);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:555]: idem

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 3 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:549]: sure. please read code.

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 2 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:540]: dm

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:552]: same

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 5 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:555]: See explanation above.

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:180]: It is not the same variable. So we can't be like you.

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 2 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:540]: Do not laugh. I used another function to handle it.

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 3 above)
Comment (hoangdat) [reply lib/features/mailbox/presentation/mailbox_controller.dart:549]: i only remind you

**Review by hoangdat — APPROVED** [review]

File: test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart [hunk 6]
```diff
@@ -0,0 +1,78 @@
+import 'package:flutter_test/flutter_test.dart';
```
Comment (hoangdat) [inline test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart:1]: file name should be end with ***_test.dart

File: test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        MailboxId(Id('A_1_1')),
+        MailboxId(Id('A_1')),
+        MailboxId(Id('A_2_1')),
+        MailboxId(Id('A_2')),
+        MailboxId(Id('A'))
+      ],
+      MailboxId(Id("B_1")): [MailboxId(Id('B_1'))],
+    };
+
+    test('_generateMapDescendantIdsAndMailboxIdList should return map mailbox id correct', () async {
```
Comment (hoangdat) [inline test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart:23]: _generateMapDescendantIdsAndMailboxIdList should return map with 2 items when mailboxes belong to 2 different tree

Comment (hoangdat) [reply test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart:23]: please add test case with 1 tree and 3 different tree also

**Review by ManhNTX — APPROVED** [review]

File: test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart [hunk 8]
```diff
@@ -0,0 +1,78 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
+import 'package:model/mailbox/presentation_mailbox.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/model/mailbox_node.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/model/mailbox_tree.dart';
+import 'package:tmail_ui_user/features/mailbox/presentation/utils/mailbox_utils.dart';
+
+void main() {
+  group('mapping delete multiple mailbox test', () {
```
Comment (hoangdat) [inline test/features/mailbox/presentation/utils/mapping_delete_multiple_mailbox.dart:10]: one group to test `map` in tuple. and the other one for list

**Review by hoangdat — APPROVED** [review]

**Pull request merged**